internal/k8s/reconciler/deployer.go

```diff
@@ -0,0 +1,121 @@
+package reconciler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/builder"
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/gatewayclient"
+	apigwv1alpha1 "github.com/hashicorp/consul-api-gateway/pkg/apis/v1alpha1"
+	"github.com/hashicorp/go-hclog"
+	apps "k8s.io/api/apps/v1"
+	core "k8s.io/api/core/v1"
+	gw "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// GatewayDeployer creates gateway deployments and services and ensures that they exist
+type GatewayDeployer struct {
+	client   gatewayclient.Client
+	consulCA string
+	sdsHost  string
+	sdsPort  int
+
+	logger hclog.Logger
+}
+
+type DeployerConfig struct {
+	ConsulCA string
+	SDSHost  string
+	SDSPort  int
+	Logger   hclog.Logger
+	Client   gatewayclient.Client
+}
+
+func NewDeployer(config DeployerConfig) *GatewayDeployer {
+	return &GatewayDeployer{
+		client:   config.Client,
+		consulCA: config.ConsulCA,
+		sdsHost:  config.SDSHost,
+		sdsPort:  config.SDSPort,
+		logger:   config.Logger,
+	}
+}
+
+func (d *GatewayDeployer) Deploy(ctx context.Context, namespace string, config apigwv1alpha1.GatewayClassConfig, gateway *gw.Gateway) error {
+	if err := d.ensureServiceAccount(ctx, config, gateway); err != nil {
+		return err
+	}
+
+	if err := d.ensureDeployment(ctx, namespace, config, gateway); err != nil {
+		return err
+	}
+
+	return d.ensureService(ctx, config, gateway)
+}
+
+func (d *GatewayDeployer) ensureServiceAccount(ctx context.Context, config apigwv1alpha1.GatewayClassConfig, gateway *gw.Gateway) error {
+	// Create service account for the gateway
+	if serviceAccount := config.ServiceAccountFor(gateway); serviceAccount != nil {
+		if err := d.client.EnsureServiceAccount(ctx, gateway, serviceAccount); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *GatewayDeployer) ensureDeployment(ctx context.Context, namespace string, config apigwv1alpha1.GatewayClassConfig, gateway *gw.Gateway) error {
+	deployment := d.Deployment(namespace, config, gateway)
```

Comment:
Should we be doing the same nil check on `deployment` here that we do for `service` in the next func?

Reply:
nope, a gateway will always have a deployment associated with it, it won't have a service unless you specifically specify a service type to use for exposing the gateway in the corresponding gatewayclassconfig